### ISSUE TITLE
Proposal: Render status view from Markdown document

### ISSUE THREAD

**Issue by csadorf** [glotzerlab/signac-flow#113]
## Proposal

The status view rendering is currently refactored to use jinja2 as part of PR #111 . After completion of that change, we should consider to split the generation and presentation of the status view into distinct _document generation_ and _document presentation_ steps.

## Current rendering process

The status view presentation is currently implemented as follows:

1. Fetch the status data.
2. Prepare status data for presentation.
3. Render status data into a terminal-presentable view (via jinja2 after merge of PR #111).

## Proposed rendering process

With the proposed change, step 3 would be split up, such that we have the following process:

1. Fetch the status data.
2. Prepare status data for presentation.
3. Render status data into a Markdown document via jinja2.
4. Render the Markdown document for presentation (either on screen or HTML).

## Advantages and Disadvantages

Such a distinct separation of document generation and presentation has multiple **advantages**:
1. It significantly simplifies the writing of a template, especially with respect to table generation.
2. The status view is independent of its final rendered presentation, *i.e.*, the view can either be presented within a terminal or in a browser (e.g. for dashboard) or something completely different (all we need is a Markdown renderer for the desired format).

**Disadvantages** are:

1. The status view generation involves one more step, which might be harder to understand for users and developers.
2. The template does not represent the exact output presented on screen.

I believe that both of these points could be addressed by providing an option to just output the "raw" document on screen, which for the most part will be identical to the rendered document (except tables).

## Example

Generating a Markdown table within a jinja2 template would look like this:
```markdown
job_id | operation name
------ | --------------
{%- for operation in operations %}
{{ operation.job_id }} | {{ operation.name }}
{%- endfor %}
```
*Note, that we do not have to worry about column widths and matching the horizontal lines with the header widths is actually purely cosmetic, `-|-` would suffice as well for valid Markdown.*

## Implementation Roadmap

For the implementation we would need to

1. Provide/utilize a Markdown renderer that renders a Markdown document into a terminal-friendly presentation, and
2. Refactor the status templates to generate a Markdown document instead of a rendered view.

Step 1 could be accomplished with one of the following alternatives:

- A patched version of [markdown_viewer](https://github.com/axiros/terminal_markdown_viewer).
- A renderer implemented with [mistune](https://github.com/lepture/mistune).
- A renderer implemented with [mistletoe](https://github.com/miyuchina/mistletoe).

While the **markdown_viewer** package would in principle provide the functionality we need, I identified a couple of issues while implementing a minimal demo for this proposal:

- Tables are not rendered correctly.
- The package is slightly over-engineered and required functionality would need to be stripped out of the package (permitted by license).
- The package has only one maintainer and package maintenance appears to be spotty.

Using **markdown_viewer** would therefore require the use of a patched version.

Alternatively, we could implement a text-renderer with either **mistune** or **mistletoe**, both packages are very popular and professionally maintained. Utilizing either of these two libraries would also allow us to render to other formats, e.g. HTML for dashboard, with the same library.

Quite frankly, my impression is that **markdown_viewer** should have been implemented through an extension of **mistune**, but maybe there were some reasons against that at the time.

I think that both **mistune** and **mistletoe** are equally viable alternatives, the former is slightly more popular, while the latter is newer. The decision which library to ultimately use -- should we decide to go this route -- could be decided during early implementation, based on which one is easier to extend for our purposes.

**Comment by csadorf** [comment]
@glotzerlab/signac-maintainers @mikemhenry @zhou-pj This is a proposal I discussed earlier with @bdice , there is no rush to decide or implemented this, but I wanted to make sure to collect thoughts and research notes.

**Comment by csadorf** [comment]
Related to issue #67 and pull request #111 .

**Comment by mattwthompson** [comment]
This seems like a good feature, but I have a couple of questions

1. Is markdown is the best way to store the intermediate data? It seems like at that step it doesn't necessarily need to be human-readable, or pretty on the eyes like markdown is. Not necessarily saying it's a bad choice or  proposing another option, just curious.

2. Is there anything preventing the user (for better or for worse) from being presented an out-of-date status? This seems to have the advantage of being cheaper - nearly free? - to view something generated by the markdown file but runs the risk of that data being out of date. Is there a cheap way of checking for this or is this meant to be the user's responsibility?

**Comment by csadorf** [comment]
> 1. Is markdown is the best way to store the intermediate data? It seems like at that step it doesn't necessarily need to be human-readable, or pretty on the eyes like markdown is. Not necessarily saying it's a bad choice or  proposing another option, just curious.

The actual status data is stored in a JSON-encodable data-structure, essentially the output of step 1.
Step 2 prepares the rendering by filtering and enriching the data and forwarding it into the jinja2 templating context.

With the proposed change we are clearly distinguishing between:
1. The status data -- that is the actual data that is available for presentation and can always be dumped in JSON-formatting using the ``--json`` option.
2. The status view document -- a _description_ on **what** we want to present and **how**.
3. The rendered status view -- the actual rendering of the status view document, i.e., the pixels on the screen if you will.

I think that the distinction between step 2 and step 3 is slightly more obvious in the context of generating a status view for the presentation in the browser, where the document would be marked up in HTML and the presented view would be the rendered HTML page.

You are correct, that we should possibly consider alternatives to which document-format we use to describe the status view document, however Markdown seems like a natural choice to me.

> 2. Is there anything preventing the user (for better or for worse) from being presented an out-of-date status? This seems to have the advantage of being cheaper - nearly free? - to view something generated by the markdown file but runs the risk of that data being out of date. Is there a cheap way of checking for this or is this meant to be the user's responsibility?

The mechanics of generating the view from a user-perspective would not really change. The moment step 1 is completed, the status is out-of-date in case that there are other simultaneous processes manipulating the data.

**Comment by mattwthompson** [comment]
Ah okay, I forgot that the status data itself is actually stored in JSON and this is all downstream from fetching the status data.

**Comment by vyasr** [comment]
I support this idea. Since we're moving towards allowing user-configured jinja templates for status view, this would make the package even more flexible from a user's perspective of how they want the status rendered.

**Comment by bdice** [comment]
@zhou-pj @csadorf This might be interesting to you as an alternative: https://github.com/willmcgugan/rich

**Comment by vyasr** [comment]
This was resolved by #163
